lib/validator.js

```diff
@@ -15,48 +54,131 @@ var inMapping = {
-            schema.required = schema.required || [];
-            schema.required.push(inMapping[param.in]);
-        }
-        var reqPartSchema = schema.properties[inMapping[param.in]];
-        reqPartSchema.properties[param.name] = {};
-        if (param.required) {
-            reqPartSchema.required = reqPartSchema.required || [];
-            reqPartSchema.required.push(param.name);
+        if (param.in !== 'body') {
+            if (!schema.properties[inMapping[param.in]]) {
+                schema.properties[inMapping[param.in]] = {
+                    type: 'object',
+                    properties: {}
+                };
+                // 'required' array must have at least one element according to json-schema spec,
+                // se we can't preinitialize it.
+                schema.required = schema.required || [];
+                if (schema.required.indexOf(inMapping[param.in]) < 0) {
+                    schema.required.push(inMapping[param.in]);
+                }
+            }
+
+            var reqPartSchema = schema.properties[inMapping[param.in]];
+            var paramSchema = { type: param.type };
+            supportedValidators.forEach(function(validator) {
+                paramSchema[validator] = param[validator];
+            });
+            reqPartSchema.properties[param.name] = paramSchema;
+            if (param.required) {
+                reqPartSchema.required = reqPartSchema.required || [];
+                reqPartSchema.required.push(param.name);
+            }
+        } else {
+            if (param.schema) {
+                schema.properties.body = param.schema;
+            } else {
+                schema.properties.body = {
+                    type: 'object'
+                };
+            }
+            if (param.required) {
+                schema.required = schema.required || [];
+                schema.required.push('body');
+            }
         }
     });
 
     return schema;
-}
+};
 
-var Validator = function(parameters) {
-    this.parameters = parameters;
-    this._validatorFunc = ajv.compile(convertToJsonSchema(parameters));
+/**
+ * Creates a function, that tries to coerce types of parameters in the
+ * incoming request, according to the provided parameters specification.
+ *
+ * @param parameters {Array} parameters swagger specification
+ * @returns {Function(req, HTTPError)} cersion function
+ * @private
+ */
+Validator.prototype._createTypeCoercionFunc = function(parameters) {
+    var code = '';
+    parameters.forEach(function(param) {
+        if (param.type === 'string') {
+            // Don't need to process strings
+            return;
+        }
+        var paramAccessor = 'req.' + inMapping[param.in] + '["' + param.name + '"]';
+        var paramCoercionCode;
+        var errorNotifier = 'throw new HTTPError({status:400,body:{type:"bad_request",'
+                + 'title:"data.' + inMapping[param.in] + '.' + param.name
```

Comment:
I'd put this as the `detail` field, while have something like *Invalid parameter specification* as the error `title`.